File: pest_control/pest_control/report/pest_trend_analysis/pest_trend_analysis.py
# Copyright (c) 2025, Meghwin Dave and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
	"""
	Execute the Pest Trend Analysis Report
	
	Args:
		filters: Dictionary of filters from the report filters
		
	Returns:
		columns: List of column definitions
		data: List of data rows
	"""
	columns = get_columns()
	data = get_data(filters)
	return columns, data


def get_columns():
	"""Define report columns"""
	return [
		{
			"fieldname": "customer",
			"label": _("Customer"),
			"fieldtype": "Link",
			"options": "Customer",
			"width": 200
		},
		{
			"fieldname": "premise_address",
			"label": _("Premise Address"),
			"fieldtype": "Link",
			"options": "Address",
			"width": 200
		},
		{
			"fieldname": "area_location",
			"label": _("Area/Location"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "pest_noticed",
			"label": _("Pest Noticed"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "total_count",
			"label": _("Total Count"),
			"fieldtype": "Int",
			"width": 100
		},
		{
			"fieldname": "occurrences",
			"label": _("Occurrences"),
			"fieldtype": "Int",
			"width": 100
		},
		{
			"fieldname": "max_activity_level",
			"label": _("Max Activity Level"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "visit_date",
			"label": _("Visit Date"),
			"fieldtype": "Date",
			"width": 120
		}
	]


def get_data(filters):
	"""Get report data based on filters"""
	conditions = get_conditions(filters)
	
	query = """
		SELECT 
			mv.customer,
			mv.customer_address as premise_address,
			ptd.area_location,
			ptd.pest_noticed,
			SUM(COALESCE(ptd.pest_count, 0)) as total_count,
			COUNT(*) as occurrences,
			MAX(ptd.activity_level) as max_activity_level,
			mv.mntc_date as visit_date
		FROM 
			`tabMaintenance Visit` mv
			INNER JOIN `tabPest Trend Detail` ptd ON ptd.parent = mv.name
		WHERE 
			mv.docstatus = 1
			{conditions}
		GROUP BY 
			mv.customer, mv.customer_address, ptd.area_location, ptd.pest_noticed, mv.mntc_date
		ORDER BY 
			ptd.pest_noticed, SUM(COALESCE(ptd.pest_count, 0)) DESC
	""".format(conditions=conditions)
	
	data = frappe.db.sql(query, filters, as_dict=1)
	return data


def get_conditions(filters):
	"""Build WHERE clause conditions"""
	conditions = []
	
	if filters.get("from_date"):
		conditions.append("mv.mntc_date >= %(from_date)s")
	
	if filters.get("to_date"):
		conditions.append("mv.mntc_date <= %(to_date)s")
	
	if filters.get("customer"):
		conditions.append("mv.customer = %(customer)s")
	
	if filters.get("premise_address"):
		conditions.append("mv.customer_address = %(premise_address)s")
	
	return " AND " + " AND ".join(conditions) if conditions else ""

File: pest_control/pest_control/report/pest_trend_analysis/pest_trend_analysis.js
// Copyright (c) 2025, Meghwin Dave and contributors
// For license information, please see license.txt

frappe.query_reports["Pest Trend Analysis"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 0
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 0
		},
		{
			"fieldname": "customer",
			"label": __("Customer"),
			"fieldtype": "Link",
			"options": "Customer",
			"reqd": 0
		},
		{
			"fieldname": "premise_address",
			"label": __("Premise Address"),
			"fieldtype": "Link",
			"options": "Address",
			"reqd": 0,
			"get_query": function() {
				return {
					"filters": {
						"link_doctype": "Customer",
						"link_name": frappe.query_report.get_filter_value("customer")
					}
				};
			}
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Highlight high activity level
		if (column.fieldname === "max_activity_level") {
			if (value === "High") {
				value = `<span style="color: red; font-weight: bold;">${value}</span>`;
			} else if (value === "Medium") {
				value = `<span style="color: orange; font-weight: bold;">${value}</span>`;
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Set default date range if not provided
		if (!report.get_filter_value("from_date") && !report.get_filter_value("to_date")) {
			report.set_filter_value("from_date", frappe.datetime.add_months(frappe.datetime.get_today(), -1));
			report.set_filter_value("to_date", frappe.datetime.get_today());
		}
	}
};

